objc_msgSend.aarch64.S

```diff
@@ -41,59 +114,114 @@
 	stp    q4, q5, [sp, #128]
+	EH_NOP
 	stp    q6, q7, [sp, #160]
-	stp    fp, lr, [sp, #192]
-	add    fp, sp, 192
-	stp    \receiver, x8, [sp, #-16]!
+	EH_NOP
+	stp    fp, lr, [sp, #192]             // The order is arbitrary, except that
+	EH_SAVE_FP_LR(192)                    // fp and lr must be spilled together
+
+	add    fp, sp, 192					  // Adjust frame pointer
+	EH_ADD_FP(192)
+	stp    x0, x8, [sp, #-16]!     // it's convenient if x0 is spilled at sp
+	EH_STACK_ALLOC(16)					  // stp performed pre-indexing by sp-16
+
+	EH_END_PROLOGUE
+
+	#ifndef _WIN32
 	.cfi_def_cfa fp, 16
 	.cfi_offset fp, -16
 	.cfi_offset lr, -8
+	#endif
 	                                      // We now have all argument registers, the link
 	                                      // register and the receiver spilled on the
 	                                      // stack, with sp containing
 	                                      // the address of the receiver
 
 	mov    x0, sp                         // &self, _cmd in arguments
-	mov    x1, \sel
-	bl     CDECL(slowMsgLookup)           // This is the only place where the CFI directives
+	mov    x1, x1
+	bl     CDECL(slowMsgLookup)           // This is the only place where the EH directives
 	                                      // have to be accurate...
 	mov    x9, x0                         // IMP -> x9
+
+	EH_START_EPILOGUE
 	ldp    x0, x1, [sp, #16]              // Reload spilled argument registers
+	EH_NOP
 	ldp    x2, x3, [sp, #32]
+	EH_NOP
 	ldp    x4, x5, [sp, #64]
+	EH_NOP
 	ldp    x6, x7, [sp, #64]
+	EH_NOP
 	ldp    q0, q1, [sp, #80]
+	EH_NOP
 	ldp    q2, q3, [sp, #112]
+	EH_NOP
 	ldp    q4, q5, [sp, #144]
+	EH_NOP
 	ldp    q6, q7, [sp, #176]
+	EH_NOP
 	ldp    fp, lr, [sp, #208]
-	ldp    \receiver, x8, [sp], #(ARGUMENT_SPILL_SIZE + 16)
+	EH_SAVE_FP_LR(208)
+
+										  // Post-increment sp += ARGUMENT_SPILL_SIZE +16
+	ldp    x0, x8, [sp], #(ARGUMENT_SPILL_SIZE + 16)
+	EH_STACK_ALLOC((ARGUMENT_SPILL_SIZE + 16))
+
+	EH_END_EPILOGUE
+	EH_END_AT_OFFSET
+
 	br     x9
 6:
-	adrp   x10, :got:SmallObjectClasses
-	ldr    x10, [x10, :got_lo12:SmallObjectClasses]
+										  // Load 63:12 of SmallObjectClasses address
+										  // We use the CDECL macro as Windows prefixes
+										  // cdecl conforming symbols with "_".
+	adrp   x10, CDECL(SmallObjectClasses) // The macro handles this transparently.
+
+										  // Add lower 12-bits of SmallObjectClasses address to x10
+	add    x10, x10, :lo12:CDECL(SmallObjectClasses)
 	ldr    x9, [x10, x9, lsl #3]
+
 	b      1b
-	.cfi_endproc
-.endm
+	EH_END
 
-.globl CDECL(objc_msgSend_fpret)
-TYPE_DIRECTIVE(CDECL(objc_msgSend_fpret), %function)
-.globl CDECL(objc_msgSend)
-TYPE_DIRECTIVE(CDECL(objc_msgSend), %function)
-.globl CDECL(objc_msgSend_stret)
-TYPE_DIRECTIVE(CDECL(objc_msgSend_stret), %function)
-CDECL(objc_msgSend):
-CDECL(objc_msgSend_fpret):
-CDECL(objc_msgSend_stret):
-	MSGSEND x0, x1
+#ifdef _WIN32
+.text
+.def objc_msgSend;
+.scl 2;
+.type 32;
+.endef
```

Comment:
Not essential for this PR, but it would be nice if these were wrapped in the same sort of macros as the ELF versions (maybe replace those with explicit `FUNCTION_DEFINE` or similar that captures what we need for both)?

Reply:
I will do this in a separate PR, as the other versions of msgSend (x86_64 and x86-32) do the same thing. We can probably streamline the function definitions in Common.S.